README.md
Xenia Mailer - FidoNet Technology Mailer
Copyright (C) 1987-2001 Arjen G. Lentz
Portions based off The-Box Mailer (C) Jac Kersing

> Xenia Mailer is free software; you can redistribute it and/or
> modify it under the terms of the GNU General Public License
> as published by the Free Software Foundation; either version 2
> of the License, or (at your option) any later version.
> 
> This program is distributed in the hope that it will be useful,
> but WITHOUT ANY WARRANTY; without even the implied warranty of
> MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
> GNU General Public License for more details.
> 
> You should have received a copy of the GNU General Public License
> along with this program; if not, write to the Free Software
> Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.



Hi everybody, these are the main Xenia mailer sources. It's a multi-platform
FidoNet-technology mailer and will currently compile under DOS and OS/2.
Jac's original code also compiled for Atari ST, and it still might.

> Jac Kersing wrote The-Box Mailer for Atari ST (and later DOS) in the 80s,
> I think even before BinkleyTerm (with sourcecode) came out. So that was a
> giant effort, no other code to refer to, only (sometimes sketchy) specs.
> While many aspects of Xenia Mailer are new, refactored or replaced, Jac
> deserves all the credit for the original architecture and making it buzz!
> We remain friends to this day.

The DOS stuff is clean for BC 3.1, you may be able to use BC 5 which is now
free for download from borland.com, but this has not been tested.
The DOS make utility used is NDMAKE 4.3 or 4.5, which is very unix-alike; don't
be using the crappy Borland make tool unless you want to rewrite the makefile.

Under OS/2 either BC/2 1.0 or Watcom C can be used, that's all stable.
Watcom offers some additional trickery, for instance the FTPMAIL facility.



Unzip this archive into a xenia\ dir under your (Xenia) development root.
To compile Xenia, you'll need to have the AGLWIN, AGLCOM and AGL_CRC packages
installed (also on https://github.com/lentzdev/)
To run Xenia, you'll need Xenia configuration files and Xenia tools.

If you're using Borland C, check TURBOC.CFG and TLINK.CFG in bc\bin\
TURBO.CFG should have lines like

	-Iu:\bc\include
	-Iu:\bc\include\sys
	-Iu:\bc\aglwin
	-Lu:\bc\lib
	-Lu:\bc\aglwin

TLINK.CFG should contain something like

	/Lu:\bc\lib;u:\bc\aglwin;


In the xenia\ directory, you will find the following stuff:

- README.md     This file
- CHANGES       The Xenia CHANGES file
- LICENSE.md    The GNU General Public License

- MAKEFILE      Generate XENIA.EXE (DOS) using Borland C, TASM and TLINK.

The make utility used is NDMAKE 4.3 or 4.5, which is very unix-alike; don't be
using the crappy Borland make tool unless you want to rewrite the makefile.

- os2/          OS/2 specific stuff. Stable.
- os2/comcapi   COMCAPI.DLL from Michael Buenter
- os2/snsrv32   Snoop server from Michael Buenter
- os2/maxdev30  MAXDEV30.ZIP (Maximus BBS 3.0 dev kit) from Lanius Corp.

All used with permission, of course.

- win32/        Start of win32 port, works without com routines so the basic
                code seems ok. But of course you need to make winnt.c tick ;)

All the DOS dependent stuff:

- exec.h        PD code by Thomas Wagner for a memory swapping spawn under DOS
- exec.c
- compat.h
- spawn.asm
- checkpat.asm  Fast wildcard pattern handling
- checkpat.h

- X00CL.OBJ     Code by Ray Gwinn to make a quicker call into X00 fossil.
- syspc.c       DOS dependent stuff for Xenia. There is some #if/#ifdef stuff
                going on in other sourcefiles too, check xenia.h for details.
                Anyway, the main thing resides here. Use as basis for porting.


The rest of Xenia Mailer:

- xenia.c      Main file - just gets things started (and terminal mode)
- xenia.h      Main include file
- proto.h      Contains all function prototype for clean ANSI compiles

- tconf.c      Configuration

- get.c        Generating manual file attachments/requests (non-windowed)
- mailer.c     Mailer - here's where the main loop is
- misc.c       Misc functions - logging, internal print, etc
- dos_file.c   Portable file I/O
- tmenu.c      Menu functions
- window.c     Window functions
- modem.c      Modem routines
- script.c     Scripting handler

- sched.h      The old but faithful event scheduler
- sched.c

- period.h     The infamous period scheduler ;)
- period.c

- nodeidx.h    Include for nodelist indexing (NODEIDX tool to create indexes)
- getnode.c    Nodelist/index read access with caching
- outblist.c   Managing the outbound directories

- mail.h       Contains defs for mail exchange (FidoNet packet formats etc)
- bbs.c        Code when starting a BBS application (spawn/exec/swap etc)
- msession.c   Mail session routines, main & incoming (mostly)
- tmailer.c    Mailer routines, outbound
- wzsend.c     WaZoo session routines (send)
- fmisc.c      Sorts out temp names for receive, recovery (Zmodem/Hydra)
- sealink.c    SEAlink protocol
- xfer.c       Modem7 filename transmit/receive routines
- cdog.c       SEAdog style file requests

- emsi.h       EMSI handshake
- emsi.c

- hydra.h      Xenia's Hydra file transfer protocol
- hydra.c

- zmodem.h     Xenia's Zmodem - this is one I (AGL) built, not a port.

This is relevant because of the improved stability/portability.

- zrecv.c      Receive code
- zsend.c      Transmit code
- zmisc.c      Stuff used by both recv and send
- zvars.c      Variables
- zcomp.c      Read this file for detail on what's not there (sorry!)

- faxproto.h   FAX receive
- faxrecv.c

- request.h    The file/service request handler (REQCOMP tool to create index)
- request.c

- md5.h        MD5 code (used for EMSI shared secret password exchange)
- md5.c



The main thing you need to do for a port is work on a syslinux.c (based
on syspc.c), creating non-blocking comport I/O (just to get a start).
See the AGLWIN package and the README.md there for ideas on how to deal with
the video/keyboard stuff.


    -- 20 May 2001, 27 January 2023
    Arjen G. Lentz



CHANGES
Xenia Mailer/Edit - log of added/changed/fixed stuff

Please note that this is NOT a complete list of changes.
Many minor things and even some major things don't get in here.
Also, it doesn't say anything about what I'm currently working on.
So don't assume too much (if anything) based on this list!


1.99.01 (OpenSource release) April/May 2001 ===================================
 - After changing work, getting engaged/married, migrating to Australia....
   You get the idea, all this took a while - life happens ;)
 - Sorting out the various bits making it workable for others
 - For XENIA project on SourceForge (sourceforge.net), see README.AGL file
 * It'd be good if someone would keep this log upto date if stuff is done ;)

1.99.00 (OpenSource preparations) 1999 ========================================

 - Because I removed the key code, Xenia now always assumes it is multiline:
   1. Now you always have to specify a commandline or config option "TASK <n>".
   2. Before, it used to delete XMRESCAN.FLG at some points if 'Single Task'.
      This made a single Xenia always do a rescan after starting, packing, etc.
      Now, it doesn't.
      If you saw Xenia 1.98.xx report itself with (Single) or (Single Task),
      check your mail process handling and touch the flag yourself if needed.
      (Yeah touch, not delete. That's cleaner.)
   3. "MAILER NO-DIRSCAN" cfg-option checked & refused if single-task.
      So NOW you can shoot yourself in the foot ;-)
   4. At some points in the code you WILL see references to serialnumbers.
      I left them in so the proper positions and structure offsets aren't lost.

 - Y2K check:
 - Fixed bug in "quoted" modem strings... the quotes were sent too ;-)
   (reported by Oliver Kopp)

1.98.06+ (GPL preparations) 16 Aug 1998 =======================================

   Xenia: Fixed dayofweek problem after 1 Mar 2000.
   Xenia/XenEdit: Cleaned up some 2-digit year prints, whenever that was
                  possible without killing formats.
   XenEdit: Cleaned up some year nasties in hudson, rbbs.

1.98.06+ (reg.only, fix/beta) 20 May 1996 =====================================

 - Start of Xenia/NT port. Well, port... most code just compiled cleanly.
   I just haven't got the windowing and async stuff sorted out, so I can't
   give you a functional EXE yet.
   HEALTH WARNING: Don't you DARE nag me about Xenia/NT unless you are an
   experienced Windows/NT programmer who can help me with the win/async code!
   (and by 'help' I mean actual writing of code, not just providing info.)

 - Fixed the remaining glitches in the period scheduler (from the .05+ alpha).
   And some other minor stuff.

 - EXTREQ statements can now take upto three of the famous %.0u to include the
   task number in the command line....

 - Commandline POLL statement now also allows polls to point-addresses.

 - Detection of orphan point .0 files, ie. things for .0 in a .PNT directory.
   For instance XenObm creates stuff like that if you poll a node with .0
   XenObm problem, but just to make sure I've added protection to Xenia....

 - Similar to above orphan story, some prog created a .D$0 which Xenia through
   some internal logic interpreted as a DLO/DUT file and started calling....
   Fixed.

 - New fax options:
   FAX PADEOL   Makes Xenia pad a fax page with EOL codes; for some view utils.
   FAX NOPADEOL The opposite, of course; this is also the default. No padding.
   FAX RAW      Select RAW fax format instead of QFX or ZFAX.
                With this option, Xenia will create a XENIAnnn.001 for the
                first page, XENIAnnn.002 for the second page, and so on.

 - LOCALNETFILE <path>
   Similar to safenetfile (passworded sessions) but LOCALNETFILE is where PKTs
   to yourself are put, like fax receive notifications. Many mailprocessors
   require this, as they don't see passworded quite the same as fully secure.
   The LOCALNETFILE setting defaults to the same path as SAFENETFILE.

1.98.05+ (restricted alpha for PERIOD scheduler) 06 April 1996 ================

 - Time on statusbar not updated during Hydra session.
   Prob created while adding the chat code. Fixed.

 - Outgoing FreePoll sometimes didn't work.
   Prob caused while adding 'MDMRSP BADCALL'. Fixed.

 - *Receiving* long filenames (OS/2 only) via Hydra didn't work. Should now.

 - DIAL statement extended with patterns! New format:
   DIAL [start] [prefix]/[suffix] [address/telnr-pattern/nodelistflags ...]
   I added this because of English ISDN implementation, where international
   ISDN data calls need to be prefixed with 000, analog calls just with 00.
   If you have an ISDN card/adapter that does both analog and ISDN, there
   is no other nice way to do it (like with prefixsuffix statements, etc).

 - Period scheduler first put to use....
   See the new XENPER.CFG for info about configuring the period scheduler.
   Various config statements will be changed to use period tags.
   INCLUDE XENPER.CFG somewhere near the start of your XENIA.CFG because
   other statements will need to use the defined period tags.....
   General rules for this:
   Multiple period tags may be specified, separated by | or + chars.
   The function is active whenever at least one period is active.
   For some functions use of period tags is optional; if no tags are
   specified then, that function will always be active.
   If no periods at all have been configured, everything is also active.

 - Event B flag (BBS) removed, changed to period logic:
     EXTAPP   <type>[,<pertags>] ...
     DIALBACK <type>[,<pertags>] ...
   This means you can control each individual extapp!
   So now could even have different BBS apps for different times ;-)
   Temporary note: the 'time to next event' parm to extapp applications
   does not work at the moment; it will always say 1440 mins (=24 hrs).

exec.c
/*
   --- Version 3.4 93-06-22 13:40 ---

   EXEC.C: EXEC function with memory swap - Prepare parameters.

   Public domain software by

	Thomas Wagner
	Ferrari electronic GmbH
	Beusselstrasse 27
	D-1000 Berlin 21
	Germany
*/

#include "compat.h"
#include "exec.h"
#include "checkpat.h"
#include <bios.h>
#include "xenia.h"

/*
   Set REDIRECT to 1 to support redirection, else to 0.
   CAUTION: The definition in 'spawn.asm' must match this definition!!
*/


#define REDIRECT  1

/*#define SWAP_FILENAME "$$AAAAAA.AAA"*/

/* internal flags for prep_swap */

#define CREAT_TEMP	0x0080
#define DONT_SWAP_ENV	0x4000

#define ERR_COMSPEC	-7
#define ERR_NOMEM	-8


spawn_check_proc *spawn_check = NULL;

/* local variables */

static char drive [MAXDRIVE], dir [MAXDIR];
static char name [MAXFILE], ext [MAXEXT];
static char cmdpath [MAXPATH] = "";
static char cmdpars [80] = "";


#ifdef __cplusplus
extern "C" int
#else
extern int _cdecl
#endif
do_spawn (int swapping,     /* swap if non-0 */
	  char *xeqfn,	    /* file to execute */
	  char *cmdtail,    /* command tail string */
	  unsigned envlen,  /* environment length */
	  char *envp	    /* environment pointer */
#if (REDIRECT)
	  ,char *rstdin,    /* redirection file names */
	  char *rstdout,
	  char *rstderr
#endif
	  );

#ifdef __cplusplus
extern "C" int
#else
extern int _cdecl
#endif
prep_swap (int method,	    /* swap method */
	   char *swapfn);   /* swap file name and/or path */

#ifdef __cplusplus
extern "C" int
#else
extern int _cdecl
#endif
exists (char *fname);

#define isslash(ch)  (ch == '\\' || ch == '/')

/* --------------------------------------------------------------------- */

/* Try '.COM', '.EXE', and '.BAT', '.BTM' on current filename,
   modify filename if found. */

static int tryext (char *fn)
{
   char *ext;

   ext = strchr (fn, '\0');
   strcpy (ext, ".COM");
   if (exists (fn))
      return 1;
   strcpy (ext, ".EXE");
   if (exists (fn))
      return 1;
   strcpy (ext, ".BAT");
   if (exists (fn))
      return 2;
   strcpy (ext, ".BTM");					/*AGL:17nov93*/
   if (exists (fn))						/*AGL:17nov93*/
      return 2; 						/*AGL:17nov93*/
   *ext = 0;
   return 0;
}

/* Try to find the file 'fn' in the current path. Modifies the filename
   accordingly. */

static int findfile (char *fn)
{
   char *path, *penv;
   char *prfx;
   int found, check, hasext;

   if (!*fn)
      return (cmdpath [0]) ? 3 : ERR_COMSPEC;

   if (isslash (fn [0]) && isslash (fn [1]))
      { 		    /* AGL: what's he upto here?? Forgot something!? */
      }
   check = checkpath (fn, INF_NODIR, drive, dir, name, ext, fn);
   if (check < 0)
      return check;

   if ((check & HAS_WILD) || !(check & HAS_FNAME))
      return ERR_FNAME;

   hasext = (check & HAS_EXT) ? 				/*AGL:17nov93*/
	    ((!stricmp (ext, ".bat") || !stricmp(ext,".btm")) ? 2 : 1) : 0;

   if (hasext)
      {
      if (check & FILE_EXISTS)
	 found = hasext;
      else
	 found = 0;
      }
   else
      found = tryext (fn);

   if (found || (check & (HAS_PATH | HAS_DRIVE)))
      return found;

   penv = getenv ("PATH");
   if (!penv)
      return 0;
   path = (char *)malloc (strlen (penv) + 1);
   if (path == NULL)
      return ERR_NOMEM;

   strcpy (path, penv);
   prfx = strtok (path, ";");

   while (!found && prfx != NULL)
      {
      while (isspace (*prfx))
	 prfx++;
      if (*prfx)
	 {
	 strcpy (fn, prfx);
	 prfx = strchr (fn, '\0');
	 prfx--;
	 if (*prfx != '\\' && *prfx != '/' && *prfx != ':')
	    {
	    *++prfx = '\\';
	    }
	 prfx++;
	 strcpy (prfx, name);
	 strcat (prfx, ext);
	 check = checkpath (fn, INF_NODIR, drive, dir, name, ext, fn);
	 if (check > 0 && (check & HAS_FNAME))
	    {
	    if (hasext)
	       {
	       if (check & FILE_EXISTS)
		  found = hasext;
	       }
	    else
	       found = tryext (fn);
	    }
	 }
      prfx = strtok (NULL, ";");
      }
   free (path);
   return found;
}


/* 
   Get name and path of the command processor via the COMSPEC 
   environmnt variable. Any parameters after the program name
   are copied and inserted into the command line.
*/

static void getcmdpath (void)
{
   char *pcmd;
   int found = 0;

   if (cmdpath [0])
      return;
   pcmd = getenv ("COMSPEC");
   if (pcmd)
      {
      strcpy (cmdpath, pcmd);
      pcmd = cmdpath;
      while (isspace (*pcmd))
	 pcmd++;
      if (NULL != (pcmd = strpbrk (pcmd, ";,=+/\"[]|<> \t")))
	 {
	 while (isspace (*pcmd))
	    *pcmd++ = 0;
	 if (strlen (pcmd) >= 79)
	    pcmd [79] = 0;
	 strcpy (cmdpars, pcmd);
	 strcat (cmdpars, " ");
	 }
      found = findfile (cmdpath);
      }
   if (!found)
      {
      cmdpars [0] = 0;
      strcpy (cmdpath, "COMMAND.COM");
      found = findfile (cmdpath);
      if (!found)						/*AGL:17nov93*/
	 {							/*AGL:17nov93*/
	 strcpy (cmdpath, "4DOS.COM");				/*AGL:17nov93*/
	 found = findfile (cmdpath);				/*AGL:17nov93*/
	 if (!found)						/*AGL:17nov93*/
	    cmdpath [0] = 0;					/*AGL:17nov93*/
	 }							/*AGL:17nov93*/
      } 							/*AGL:17nov93*/
}


/*
   tempdir: Set temporary file path.
	    Read "TMP/TEMP" environment. If empty or invalid, clear path.
	    If TEMP is drive or drive+backslash only, return TEMP.
	    Otherwise check if given path is a valid directory.
	    If so, add a backslash, else clear path.
*/

static int tempdir (char *outfn)
{
   int i, res;
   char *stmp [4];

   stmp [0] = getenv ("TMP");
   stmp [1] = getenv ("TEMP");
   stmp [2] = ".\\";
   stmp [3] = "\\";

   for (i = 0; i < 4; i++)
      if (stmp [i])
	 {
	 strcpy (outfn, stmp [i]);
	 res = checkpath (outfn, 0, drive, dir, name, ext, outfn);
	 if (res > 0 && (res & IS_DIR) && !(res & IS_READ_ONLY))
	    return 1;
	 }
   return 0;
}


#if (REDIRECT)

static int redirect (char *par, char **rstdin, char **rstdout, char **rstderr)
{
   char ch, sav;
   char *fn, *fnp, *beg;
   int app;

   do
      {
      app = 0;
      beg = par;
      ch = *par++;
      if (ch != '<')
	 {
	 if (*par == '&')
	    {
	    ch = '&';
	    par++;
	    }
	 if (*par == '>')
	    {
	    app = 1;
	    par++;
	    }
	 }

      while (isspace (*par))
	 par++;
      fn = par;
      if ((fnp = strpbrk (par, ";,=+/\"[]|<> \t")) != NULL)
	 par = fnp;
      else
	 par = strchr (par, '\0');
      sav = *par;
      *par = 0;

      if (!strlen (fn))
	 return 0;
      fnp = (char *)malloc (strlen (fn) + app + 1);
      if (fnp == NULL)
	 return 0;
      if (app)
	 {
	 strcpy (fnp, ">");
	 strcat (fnp, fn);
	 }
      else
	 strcpy (fnp, fn);

      switch (ch)
	 {
	 case '<':   if (*rstdin != NULL)
			return 0;
		     *rstdin = fnp;
		     break;
	 case '>':   if (*rstdout != NULL)
			return 0;
		     *rstdout = fnp;
		     break;
	 case '&':   if (*rstderr != NULL)
			return 0;
		     *rstderr = fnp;
		     break;
	 }

      *par = sav;
      strcpy (beg, par);
      par = strpbrk (beg, "<>");
      }
   while (par);

   return 1;
}

#endif


int do_exec (char *exfn, char *epars, int spwn, unsigned needed, char **envp)
{
   static char swapfn [MAXPATH];
   static char execfn [MAXPATH];
   unsigned avail;
   union REGS regs;
   unsigned envlen;
   int rc, ffrc;
   int idx;
   char **env;
   char *ep, *envptr, *envbuf;
   char *progpars;
   int swapping;
#if (REDIRECT)
   char *rstdin = NULL, *rstdout = NULL, *rstderr = NULL;
#endif

   envlen = 0;
   envptr = NULL;
   envbuf = NULL;

   if (epars == NULL)
      epars = "";
   if (exfn == NULL)
      execfn [0] = 0;
   else
      strcpy (execfn, exfn);

   getcmdpath ();

   /* First, check if the file to execute exists. */

   if ((ffrc = findfile (execfn)) <= 0)
      return RC_NOFILE | -ffrc;

   if (ffrc > 1)   /* COMMAND.COM or Batch file */
      {
      if (!cmdpath [0])
	 return RC_NOFILE | -ERR_COMSPEC;

      idx = (ffrc == 2) ? strlen (execfn) + 5 : 1;
      progpars = (char *)malloc (strlen (epars) + strlen (cmdpars) + idx);
      if (progpars == NULL)
	 return RC_NOFILE | -ERR_NOMEM;
      strcpy (progpars, cmdpars);
      if (ffrc == 2)
	 {
	 strcat (progpars, "/c ");
	 strcat (progpars, execfn);
	 strcat (progpars, " ");
	 }
      strcat (progpars, epars);
      strcpy (execfn, cmdpath);
      }
   else
      {
      progpars = (char *)malloc (strlen (epars) + 1);
      if (progpars == NULL)
	 return RC_NOFILE | -ERR_NOMEM;
      strcpy (progpars, epars);
      }

#if (REDIRECT)
   if ((ep = strpbrk (progpars, "<>")) != NULL)
      if (!redirect (ep, &rstdin, &rstdout, &rstderr))
	 {
	 rc = RC_REDIRERR;
	 goto exit;
	 }
#endif

   /* Now create a copy of the environment if the user wants it. */

   if (envp != NULL)
      for (env = envp; *env != NULL; env++)
	 envlen += strlen (*env) + 1;

   if (envlen)
      {
      /* round up to paragraph, and alloc another paragraph leeway */
      envlen = (envlen + 32) & 0xfff0;
      envbuf = (char *)malloc (envlen);
      if (envbuf == NULL)
	 {
	 rc = RC_ENVERR;
	 goto exit;
	 }

      /* align to paragraph */
      envptr = envbuf;
      if (FP_OFF (envptr) & 0x0f)
	 envptr += 16 - (FP_OFF (envptr) & 0x0f);
      ep = envptr;

      for (env = envp; *env != NULL; env++)
	 {
	 ep = stpcpy (ep, *env) + 1;
	 }
      *ep = 0;
      }

   if (!spwn)
      swapping = -1;
   else
      {
      /* Determine amount of free memory */

      regs.x.ax = 0x4800;
      regs.x.bx = 0xffff;
      intdos (&regs, &regs);
      avail = regs.x.bx;

      /* No swapping if available memory > needed */

      if (needed < avail)
	 swapping = 0;
      else
	 {
	 /* Swapping necessary, use 'TMP' or 'TEMP' environment variable
	   to determine swap file path if defined. */

	 swapping = spwn;
	 if (spwn & USE_FILE)
	    {
	    if (!tempdir (swapfn))
	       {
	       spwn &= ~USE_FILE;
	       swapping = spwn;
	       }
	    else if (OS_MAJOR >= 3)
	       swapping |= CREAT_TEMP;
	    else
	       {
	       /*strcat (swapfn, SWAP_FILENAME);*/
	       sprint(&swapfn[strlen(swapfn)],"XENIA%02x$.SWP",task);
	       idx = strlen (swapfn) - 1;
	       while (exists (swapfn))
		  {
		  if (swapfn [idx] == 'Z')
		     idx--;
		  if (swapfn [idx] == '.')
		     idx--;
		  swapfn [idx]++;
		  }
	       }
	    }
	 }
      }

   /* All set up, ready to go. */

   if (swapping > 0)
      {
      if (!envlen)
	 swapping |= DONT_SWAP_ENV;

      rc = prep_swap (swapping, swapfn);
      if (rc < 0)
	 rc = RC_PREPERR | -rc;
      else
	 rc = 0;
      }
   else
      rc = 0;

   if (!rc)
      {
      if (spawn_check != NULL)
	 rc = spawn_check (ffrc, swapping, execfn, progpars);
      if (!rc)
#if (REDIRECT)
	 rc = do_spawn (swapping, execfn, progpars, envlen, envptr, rstdin, rstdout, rstderr);
#else
	 rc = do_spawn (swapping, execfn, progpars, envlen, envptr);
#endif
      }

   /* Free the environment buffer if it was allocated. */

exit:
   free (progpars);
#if (REDIRECT)
   if (rstdin)
      free (rstdin);
   if (rstdout)
      free (rstdout);
   if (rstderr)
      free (rstderr);
#endif
   if (envlen)
      free (envbuf);

   return rc;
}

